### ISSUE TITLE
Custom Classes proposal

### ISSUE THREAD

**Issue by x87** [sannybuilder/dev#37]
Here is the proposal to add a new syntax for defining custom classes in scripts. They may eventually become a replacement for static classes in the classes.db.

Let's take a look at an example:

```
class Actor
  putInCar(car)
     036A: put_actor self in_car car
  end
end

var
  0@: Actor
end

0@.putInCar($car)
```

Here we defined a class named `Actor` that has one method `putInCar` in. `putInCar` method has one parameter `car`. Then we declared a variable `0@` of the type `Actor`  and finally called method `putInCar` on that class instance.

### Important Notes

* class declaration is lazy. it means no code is produced until it gets actually used in the code. if we were to remove `0@.putInCar($car)` no code was produced in this script.
* each method receives an implicit variable `self` that represents the class instance the method is called on. in this case it is the variable `0@`
* in this example  `0@.putInCar($car)` gets compiled into `036A: put_actor 0@ in_car $car`

## Design Choices
* when there is a method call statement the compiler takes the method body and replaces the statement with the entire body. it means the method could have many instructions and possibly even labels and they will be put in place of the call statement. in this sense class method are more templates than functions as in the other languages.
*  this however does not prevent local variables from being altered inside the method body. it means there is no separate scope created when the class method is called and if you, for example, change the variable 0@ in the class method it will be a side-effect for the callee:

```
class Actor
  putInCar(car)
     036A: put_actor self in_car car
     0@ = 2
  end
end

0@ = 1
0@.putInCar($car)
// 0@ is now equal to 2
```

There are 3 possible choices to this dilemma:

1) leave it as is, considering this a liberate language design decision (when you invoke subroutines with `gosub` they could alter local scope as well)
2) forbid using local variables syntax inside methods. only use arguments passed into the method. this could be inconvenient if you need to keep a local state.
3) add extra prologue and epilogue before each method call that will store and restore all local variables. this adds extra overhead in runtime

**Comment by x87** [comment]
method could be conditional:
```
class Actor
  inAnyCar()
     00DF:   actor self in_any_car 
  end
end
```
```

```
var
  0@: Actor
end
if
  0@.InAnyCar()
then
```
gets compiled as

```
if
   00DF:   actor 0@ in_any_car
then
```

**Comment by x87** [comment]
It's unclear how to work with returning a value from the methods, i.e consider:

```
class Actor
  pos()
     00A0: store_actor self position_to x y z
  end
```

how to represent `x`, `y`, `z` in the code?

## Design Choices

1) pass variable into the method as parameters
```
  pos(x, y, z)
     00A0: store_actor self position_to x y z
  end
```

```
0@.pos(1@, 2@, 3@) // 00A0: store_actor 0@ position_to 1@ 2@ 3@
```

2) add an implicit variable `result` then will represent the left side of the call expression:

```
  pos()
     00A0: store_actor self position_to result[0] result[1] result[2] // variant 1
     00A0: store_actor self position_to result.0 result.1 result.2 // variant 2
  end
```
then
```
( 1@, 2@, 3@ ) = 0@.pos() // 00A0: store_actor 0@ position_to 1@ 2@ 3@
```

this adds a lot of unusual syntax into the language.

**Comment by wmysterio** [comment]
Мощно! Вопрос: будет ли подхватывать компилятор декларированные классы, если они описаны в другом файле и были подключены через директиву {$i}. Есть планы написать такие классы (у меня уже есть всё готовое для этого) и хотелось бы избавить разработчика от копирование всего каждый раз.

**Comment by x87** [comment]
@wmysterio 
yes, the compiler definitely should be able to read class declarations from imported files. 

static class variables could be a nice feature! Will keep that in mind.

**Comment by x87** [comment]
Class instances should support inline declaration similar to primitive types:

```
int x = 0
Player p = Player.Create(x, 10.0, 20.0, -100.0)
```

**Comment by wmysterio** [comment]
Is it possible to support optional parameters?
```delphi
class Actor
    is_near_point_3d( x, y, z, drawSphere = 1, radiusX = 1.0, radiusY = 1.0 radiusZ = 1.0 )
    00FE:      self  drawSphere   x   y   z   radiusX   radiusY   radiusZ
    end
end

//...
//if
$hActor.is_near_point_3d( -1244.0, 123.0, 16.0 )
$hActor.is_near_point_3d( -1244.0, 123.0, 16.0, 1 )
$hActor.is_near_point_3d( -1244.0, 123.0, 16.0, 1, 1.25, 1.25, 1.25 )
$hActor.is_near_point_3d( -1244.0, 123.0, 16.0, radiusZ=2.0 )
```

> how to represent x, y, z in the code?

Maybe need use additional syntax:
````
00A0: store_actor {self} position_to {x} {y} {z} // Conflict with comments syntax. Can be fixed specifically for the class
00A0: store_actor (self) position_to (x) (y) (z)
````

**Comment by x87** [comment]
2024 Draft
```

class Car
    /// Directly mapped to command CREATE_CAR
    /// Function and Command signatures must match
    Create<0x00A5>(modelId: int, x: float, y: float, z: float): Car

    /// Update is an instance method
    Update(self: Car, flag: int)
        // update
    end

    /// Can return logical values
    Check(self: Car): logical
        return true
    end
end

Car c = Car.Create(#INFERNUS, 0, 0, 0) // compiled as 00A5
if c.Check() // compiled as a scm function
then
    c.Update(1) // compiled as a scm function
end
```
